pkg/util/fsm/fsm.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2017 ScyllaDB
+
+package fsm
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// ErrEventRejected is the error returned when the state machine cannot process
+// an event in the state that it is in.
+var ErrEventRejected = errors.New("event rejected")
+
+const (
+	// Default represents the default state of the system.
+	Default StateType = ""
+
+	// NoOp represents a no-op event.
+	NoOp EventType = "NoOp"
+)
+
+// StateType represents an extensible state type in the state machine.
+type StateType string
+
+// EventType represents an extensible event type in the state machine.
+type EventType string
+
+// EventContext represents the context to be passed to the action implementation.
+type EventContext interface{}
+
+// Action represents the action to be executed in a given state.
+type Action func(ctx context.Context) (EventType, error)
+
+// Events represents a mapping of events and states.
+type Events map[EventType]StateType
+
+// State binds a state with an action and a set of events it can handle.
+type State struct {
+	Action Action
+	Events Events
+}
+
+type Hook func(ctx context.Context, currentState, nextState StateType, event EventType) error
+
+// States represents a mapping of states and their implementations.
+type States map[StateType]State
+
+// StateMachine represents the state machine.
+type StateMachine struct {
```

Comment:
You are exporting mutex protected fields directly, I'd suggest to use builder or at least a constructor.